### ISSUE TITLE
Migration process: Autosubmit API to Python 3.7.3

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#446]
In GitLab by @wuruchi on Nov 15, 2019, 09:25

Hello @dbeltrankyl @mcastril 

In this issue, I am going to document the process of migrating the Autosubmit API to Python 3.7.3. That includes the main method(s) and tools I use; also, I am going to list the main errors I encounter and their details. All of this with the purpose of helping with the migration of Autosubmit.

Currently, the API lives as a branch of Autosubmit, but it also has its own PyPi repository: [https://pypi.org/project/autosubmitAPIwu/](https://pypi.org/project/autosubmitAPIwu/)

A new PyPi repository will be created once the migration is completed.

I am going to use **Python/3.7.3-foss-2015a** in a local environment.

Updates will follow.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 15, 2019, 09:25

changed due date to November 22, 2019

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 18, 2019, 15:59

Hello @mcastril, @dbeltrankyl 

### First attempt report:

These are the steps that have been taken to migrate the API to Python3:

1. Bscearth.utils has been migrated to Python 3.7.3. Here is the new repository for that version: [https://earth.bsc.es/gitlab/ces/bscearth.utils3](https://earth.bsc.es/gitlab/ces/bscearth.utils3). Also, the PyPi repository has been created: [https://pypi.org/project/bscearth.utils3/](https://pypi.org/project/bscearth.utils3/). See further points for more information on how important this is.
2. Using [2to3](https://docs.python.org/2/library/2to3.html), I tried to adapt as much code as possible before trying to compile Autosubmit in the new version. This worked without serious problems.

3. Using [caniusepython3](https://pypi.org/project/caniusepython3/) I checked dependencies and there were not any notable exceptions or warnings found. It reported that a couple of libraries needed to be updated: a Flask related library, and saga-python. So, saga-python has been replaced by radical.saga, and the Flask related library was installed without a problem.

4. I created a virtual environment for Python 3.7.3 where I installed all the required dependencies for Autosubmit. During this process, I noted that it is possible to use the current version of bscearth.utils as a reference for this Python 3.7.3 project. After I finished installing all the required dependencies I tried to compile `autosubmit.py`.

5. No significant problems were found during compilation. Some references were broken but could be easily fixed. After that, I started the flask server to test the API functionality. There were some missing references, but solving them was trivial.

6. The first problem that I encountered is that the `pkl` file is not saved in the same format in Python 3.7.3, and as a consequence, it needs extra parameters to be read. Basically, to read a Python 2.7 pkl file we need to add `encoding=bytes` so the line that reads the pkl in `job_list_persistence.py` has the following declaration `pickle.load(fd, encoding='bytes')`. I suppose that something similar should be done to the `save()` function. Since the API does not save any pkl, I have not tested that.

7. Then, while testing the Graph generation function everything seemed to work well until I tried to visualize an experiment with wrappers. Here is where I found an exception related to `pydotplus` that basically is caused by some restrictions that Python 3.7.3 imposes on how dictionaries are handled. This seems to be a problem of the library, which is targeted mainly for Python 2.7 and 3.4. **This is an application breaking problem.** I will keep looking into it and try to find a workaround. The last release of pydotplus was on December 9th, 2014.

8. *UPDATE*: `pydot` seems to be a good replacement for `pydotplus`, more testing is needed.

I created a repository for the Python3 version of AutosubmitAPI: [https://earth.bsc.es/gitlab/ces/autosubmitapi3](https://earth.bsc.es/gitlab/ces/autosubmitapi3). The Python2 version still lives and will continue living in the branch **issue363** of Autosubmit.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 21, 2019, 08:55

changed due date to November 29, 2019

**Comment by kinow** [comment]
In GitLab by @mcastril on Nov 21, 2019, 19:56

Thanks for the log @wuruchi , it's very detailed and helpful!

I have a couple of questions only:

> During this process, I noted that it is possible to use the current version of bscearth.utils as a reference for this Python 3.7.3 project. 

What do you mean with this?

> Here is where I found an exception related to `pydotplus` that basically is caused by some restrictions that Python 3.7.3 imposes on how dictionaries are handled. 

And I guess we cannot use standard pydot ?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 22, 2019, 09:17

Hello @mcastril 

>What do you mean with this?

Under Python 3.7.3 I executed `pip install bscearth.utils` and it worked just fine. Also, when compiling and running under Python 3.7.3, the compiler didn't complain.

>And I guess we cannot use standard pydot ?

It seems like no so far. I haven't tested any further because I was working on the grouping function of the graph, but I will give it a couple of hours today. In the worst-case scenario, we can fork the repository and make our own version of that library, given that it will be almost the same as it is now but with a couple of fixes for 3.7.3.

**Comment by kinow** [comment]
In GitLab by @mcastril on Nov 22, 2019, 10:21

> Under Python 3.7.3 I executed `pip install bscearth.utils` and it worked just fine. Also, when compiling and running under Python 3.7.3, the compiler didn't complain. 

So you migrated bscearth.utils to Python3 in bscearth.utils3, but you still can use bscearth.utils in a Python3 project... 

> In the worst-case scenario, we can fork the repository and make our own version of that library, given that it will be almost the same as it is now but with a couple of fixes for 3.7.3. 

Yes, that was also the first thing I thought about, but maybe is better to see if we can use another one that is externally maintained. I think I have used Graphviz with pydot, but maybe we need special features.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 22, 2019, 10:25

>So you migrated bscearth.utils to Python3 in bscearth.utils3, but you still can use bscearth.utils in a Python3 project...

I migrated bscearth.utils to Python3 in bscearth.utils3 but ultimately I didn't need to use that, I just installed bscearth.utils using pip. So, there is a Python3 version that I migrated that I am not using right now because the Python2 version seems to be working just fine under Python3.

>Yes, that was also the first thing I thought about, but maybe is better to see if we can use another one that is externally maintained. I think I have used Graphviz with pydot, but maybe we need special features.

Sure, before attempting to fork the current pydotplus we can try other libraries that perhaps offer better performance.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 22, 2019, 13:07

Hi,

watching the pipeline I noticed that it may begin to fail on  **January 1st, 2020** for all current code.

report codacy step:
```
$ pip install codacy-coverage --upgrade
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
```


```
    WARNING: saga-python is deprecated!

    Your application will still work - but "saga-python" has been renamed to
    "radical.saga".  Please change your imports from

        import saga

    to

        import radical.saga as saga

    No other changes are needed.  This backward compatibility package will disappear in version 1.0, planned for the end 2019.
```

Apparently pip won't work with 2.7, and about saga it doesn't work anyway we should maintain it on 3.x?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 22, 2019, 15:23

Hi @mcastril 

Update on `pydotplus`: `pydot` seems to be a good replacement. After replacing `pydotplus.graphviz.Cluster` by `pydot.Cluster` it works.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 28, 2019, 09:06

removed due date

**Comment by kinow** [comment]
In GitLab by @wuruchi on Nov 28, 2019, 09:06

changed due date to December 06, 2019

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 2, 2019, 10:59

Hi @dbeltrankyl . I guess it will work unless we upgrade Pip right? So we shouldn't upgrade and continue using the same version. Could we freeze pip for some time? I don't know which is the strategy for that. @kserradell ?

If we don't use Saga anymore I think we should drop it.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 2, 2019, 11:59

Pip will work with Python2.7 but won't be maintained.

Is there any AS branch to test installation of AS with Python3?

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 2, 2019, 13:19

Hi @kserradell . We know there will be issues with some libraries that Autosubmit is using (some of them developed by ourselves). So we decided to start porting the Autosubmit API (the one used by the GUI and that has many of the functionalities of Autosubmit itself). 

It seems @wuruchi made it work after addressing some issues (described at the top), so after he confirms that we will start working on Autosubmit.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 2, 2019, 13:35

Hello @mcastril @kserradell

Since I kept working on the Python 2 version of the Autosubmit API, and as a consequence the Python 3 version is behind, I will make another migration attempt this week including all the fixes found in the first attempt. I am going to update this issue with another report as soon as the migration is completed.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 3, 2019, 09:43

Ok, so please keep me informed when there is a full AS version with Python3 support able to be deployed.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 5, 2019, 10:12

Thanks for the update @wuruchi!

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 5, 2019, 13:43

Hello @mcastril @dbeltrankyl @kserradell 

The migration of AutosubmitAPI to **Python/3.7.3-foss-2015a** has been completed successfully (as far as my testing can tell). 

## Second and Final Attempt report:

The migration process has left these important remarks.

1. It was not necessary to use the *Bscearth.utils* Python3 version that I ported, although the API only uses a couple of its functionalities. Executing `pip install Bsceart.hutils` from the Python3 virtual environment worked. I guess the current version of *Bscearth.utils* is fully compatible with Python3, time will tell.

2. Executing `2to3 -w script.py` where "script" is the name of your python script seems to have solved most of the possible syntax errors. Furthermore, this command (it is a pip package) generates a backup file for each of the migrated scripts.

3. The workaround to fully read pkl files generated with Python2 includes the code: `with open(path, 'rb') as fd: return pickle.load(fd, encoding='latin1') ` where before I suggested `encoding='bytes'` that indeed returns a list of jobs, but which inner data cannot be properly read, `'latin1'` solves that issue. As for saving the pkl, the recommended code is `pickle.dump()`, which is already implemented.

4. The compiler complained about comparisons between sets/lists and integers, easily solved.

5. There is a problem with the "configparser" from Python3, it happened with the experiment a2dw, when the parser found in the "proj.conf" section `[save_ic]` the parameter `SAVE_IC_COND` with value `(( leg_number%5 == 0 ))` it threw a syntax error. Apparently, `%` is some sort of identifier in this version of the library. It was solved by ignoring "proj.conf" check, but that only works for the API, Autosubmit needs that information.

6. There is another less critical problem with "configparser", it throws an error when a parameter is defined twice in the same section. This is logical, but in the Python2 version it did not happen.

7. The `pydot` fixes work. Seems to be faster, but that might be wishful thinking.

8. Since most of the errors in Autosubmit are properly handled, it could be useful to use `traceback`, a python library that allows printing the traceback of an exception in a controlled way. I mean, for debugging purposes once we start to work in the migration of Autosubmit.

## Next Steps

The installation of Python 3.7.3 in bscesweb04 has been requested, which includes python-devel. Once that is completed, the Python3 version of Autosubmit API will be deployed in bscesweb04. 

Work on the Python2 version of Autosubmit API will be stopped. Only Python3 version from now on. The API now lives in [https://earth.bsc.es/gitlab/ces/autosubmitapi3](https://earth.bsc.es/gitlab/ces/autosubmitapi3).

Try migrating Autosubmit.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 5, 2019, 15:18

Thanks @wuruchi. The installation of Python 3.7.3 in bscesweb04 including python-devel have been done.

@dbeltrankyl would it be possible for you to start the migration?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 5, 2019, 15:23

Hi @kserradell , In autosubmit meeting we decided to stop developments once I update the last changes that should be ready in 1hour~, 

From next week I'll be focused on migrating to python 3.7.3

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 5, 2019, 16:08

Thanks for the info. Sounds good.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 9, 2019, 10:31

removed due date

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 9, 2019, 11:16

Hello @kserradell 

In bscesweb04, I could only find Python 3.4.6.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 9, 2019, 12:54

My mistake @wuruchi. Unfortunately, this is the version available in the SLES12 repos. Can you use this version?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 9, 2019, 13:03

Hello @kserradell 

Sure, I will try to make it work with that version. According to the documentation, Python 3.7 code is compatible with 3.4 code, so there should not be any major problem.

I will update the issue when it is completed.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 9, 2019, 14:50

Thanks. This machine is administrated by Operations and they don't like to add non-official repos.

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 9, 2019, 19:25

Sorry to give the negative view here, but Python 3.4 is already deprecated, as Python 2 will be. I guess that's why you want to update the current image? Or maybe I am wrong. Well, the issue is that we should use at least 3.6 if we can tell this to Operations.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 10, 2019, 09:16

Hello @mcastril 

As you mention, Python 3.4 has been deprecated for some months now, but we can use it for the API in the meanwhile, hoping that Operations upgrades it soon enough. Anyway, it will be great to have a current version like Python 3.6 at least.

Also, Python 3.4 does not have any *python-saga* or *radical.saga* implementation, so I had to delete all the references to that from the API code, I suppose it would not be easy with the actual Autosubmit code.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Dec 10, 2019, 09:18

Adding to @mcastril message. Can we agree on which version of Python we are going to port Autosubmit? I think Python 3.7.3 would be ideal, but I guess it also depends on which version is available in the machines where it will run.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 10, 2019, 10:22

bscesweb04 is a production webserver and will not be updated, at least in the next months. I will ask to look for a repo with python 3.7.*.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 13, 2019, 17:34

Congratulations @dbeltrankyl !!!

About modules not showing, is due to cache issues. Execute `rm -r ~/.lmod` and run a `module av` again.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 16, 2019, 11:59

mentioned in commit bae9f92c1a62d336bbf2f8f71ba5d07811946b9b

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 16, 2019, 14:50

mentioned in commit 5c24f9da2e43c92db770169465b0cba69f9ae690

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 16, 2019, 15:21

Hello, 

I fixed two issues related to bytes handling and the tests. The problem is that for python 3 MOCK object acts as a string and when the real code does that work is using bytes. For example, opening a file in the real code the text loaded output would be in bytes type while if it is open with MOCK it will be loaded as String type. So it crashed there and reported a false negative, now it is solved. 

I fixed another issue related with the pickle with the help of @wuruchi, the problem resided in that python 3 uses a new protocol for store the data ( and load but that is not a problem with the encoding='latin1') which is incompatible with python 2.9, so now the pickle store in python 3 is changed to use an older protocol compatible with python 2 but, I'm not sure if there will be implications in speed, security... Maybe some more compatibility issues will raise when combining python 2 and python 3 experiments with a mix of python 2 GUI with python 3 autosubmit or/and vice-versa, especially I'm thinking when interactions with the GUI will be implemented ( setstatus, security...).


On a side note, I performed more tests and 4.0b appears to be working fine, @mcastril can auto-earth work with python 3 for performing some weekly test? or some model just to ensure that the content of the templates will be the same as in python 2.


@mcastril @wuruchi, additionally I don't know if there will be any issues if a running experiment decides to change the version to 4.0 in a middle of a run maybe we should implement an error if a user decides to use 4.0 for a 3.X version, displaying "Your current experiment version is incompatible with this autosubmit version, if you want to continue please do autosubmit create + autosubmit recovery?" (or some like that)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 16, 2019, 15:26

In a slightly different topic, @wuruchi @mcastril 

Now that the autosubmit code is ported to python 3, I'm looking forward to migrate the template code


What we should do with the code inside the textwrap that will be included on the templates as a header. Currently, they're still on python 2 but should autosubmit have an extra parameter for a while to be able to send python 2 header and python 3 header for model compatibility possible issues? 

For example:

```
 def get_nodes(self):
        return textwrap.dedent("""
        # Getting the list of allocated nodes
        {0}
        os.system("mkdir -p machinefiles")

        with open('node_list', 'r') as file:
             all_nodes = file.read()

        all_nodes = all_nodes.split("_NEWLINE_")
        """).format(self.allocated_nodes, '\n'.ljust(13))
```

this would crash in a python 3 model.

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 16, 2019, 18:59

Good idea, we have to take in mind this. I think we can expose an option in autosubmit.conf to be able to work with one or the other version in the remote machines.

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 16, 2019, 19:03

Congratulations @dbeltrankyl ! Very good job and very efficient one.

About the modules, I am a somehow afraid about exposing it to everybody. This is not even an alpha version and I think we should clearly specify it. 

Could you change the name 

`autosubmit/4.0b-foss-2015a-Python-3.7.3`

To something like 

`autosubmit/4.0-pre-alpha-foss-2015a-Python-3.7.3`

so it is totally clear?

With 4.0b it would be at the same testing level as 3.12.0b which is not the case.

Thanks,

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 16, 2019, 19:11

My opinion is that it is a little bit early for that... We will move the default version to `3.12.0b` which has been used in the weekly testing suite for a while. At that moment the weekly suite will start using `3.12.1b` which is now our development version but that we have been using for some time and it is kind of robust.

In parallel Pablo is testing 3.12.1b remote dependencies on a copy of a real experiment.

I think we should test 4.0 a little bit more ourselves before the step of asking others' to do it.

@dbeltrankyl , I guess you already completed a dummy experiment with Autosubmit 4.0b. Could you copy and run the simplest Auto-EC-Earth experiment from the weekly tests? For example, experiment t01q from the last working tag:

https://earth.bsc.es/gitlab/es/auto-ecearth3/-/tags

Config files are in a submodule since that tag, so you should use that ones in the test_cases folder.

Once that we can technically run a simple experiment with this version I think we can ask some users to run the same but verifying everything is correct.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 10:00

Hi, that makes more sense I will change it, how I delete an already pushed module then? `autosubmit/4.0b-foss-2015a-Python-3.7.3` (I don't want to mess it up).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 10:05

Hi @mcastril,

>  @dbeltrankyl , I guess you already completed a dummy experiment with Autosubmit 4.0b. 

Yes I completed several dummy experiments ( with wrappers and without) 

> Could you copy and run the simplest Auto-EC-Earth experiment from the weekly tests? For example, experiment t01q from the last working tag ...

I tried to do this and it is crashing in local_setup, probably because I have python 3 module loaded and the local setup template uses a code written in python 2:

/esarchive/autosubmit/a2gg/tmp/LOG_a2gg/a2gg_LOCAL_SETUP.20191217095539.err
```
+ python3 /esarchive/autosubmit/a2gg/proj/auto-ecearth3/plugins/config_validator/main.py t01q /esarchive/autosubmit/a2gg/conf
Traceback (most recent call last):
  File "/esarchive/autosubmit/a2gg/proj/auto-ecearth3/plugins/config_validator/main.py", line 21, in <module>
    exit(main())
  File "/esarchive/autosubmit/a2gg/proj/auto-ecearth3/plugins/config_validator/main.py", line 15, in main
    ret = config_validator.validate()
  File "/esarchive/autosubmit/a2gg/proj/auto-ecearth3/plugins/config_validator/configValidator.py", line 37, in validate
    ret = TemplateName(self.variables["proj"]["common"]['template_name'], self.variables).validate() and \
KeyError: 'common'
 self.variables).validate() and \
KeyError: 'common'
```


I guess the issue could be solved specifying the python version inside the code instead of relying on ENV variables 

EDIT1: nvm, I can't run it on 3.12 

Edit 2: I can't run t026 neither t01q it breaks in the same step 

EDIT 3: I needed to change proj.conf, VERSION = 3.3.1_trunk_week_2019_48 I only changed it on expdef now it works on 3.12.0 gonna check 4.0

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 10:12

In relation to the other thread. 

>> Good idea, we have to take in mind this. I think we can expose an option in autosubmit.conf to be able to work with one or the other version in the remote machines.

Perhaps should I  expose an option to jobs.conf instead of autosubmit.conf ( or in both locations) to select if the script template is using python2? some like

**autosubmit.conf**
```
[config]
(...)
PYTHON2 = False # default False
```

**jobs.conf**
```
[LOCAL_SETUP]
(...)
PYTHON2 = True  # default False
```

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 17, 2019, 10:53

I can delete the module for you. Please confirm.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 10:55

Thanks, @kserradell do it, please.

**Comment by kinow** [comment]
In GitLab by @kserradell on Dec 17, 2019, 10:59

Done.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 12:28

The last error was a mistake for my side, I didn't know that I must change also the branch in proj.conf ( VERSION = TRUNK TO VERSION = 3\.3.1_trunk_week_2019_48).

However, it raises the error I was mentioning (now for real)

```
/esarchive/autosubmit/a2gg/proj/auto- ecearth3/plugins/check_procesors_configuration.py a2gg
Traceback (most recent call last):
  File "/shared/earth/software/Python/2.7.9-foss-2015a/lib/python2.7/encodings/__init__.py", line 74, in search_function
    entry = _cache.get(encoding, _unknown)
AttributeError: 'NoneType' object has no attribute 'get'
ibuteError: 'NoneType' object has no attribute 'get'
```


While local.setup is capable of use python 3 and python 2 together, it only works assuming that autosubmit is in python 2

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 12:59

Tweaking `local_setup` in local ( just to see what is wrong and see how to translate it to autosubmit), I see that the issue is that even if you perform `module purge`, `PYTHONPATH` bash variable remains with the info of 3.7.3, performing  `module load python/2.7.9-foss-2015a` doesn't change that fact. The solution for the model is just to perform `unset PYTHONPATH` whenever wants to use python2. But I'll look to do that globally in autosubmit for python 2 models

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 17, 2019, 13:03

Sidenote: I noticed that 

```
function check_procesors_configuration() {
    set +xuve
    module purge
    module load bscearth.utils/0.5.2-foss-2015a-Python-2.7.9
    module load autosubmit
    set -xuve

    $PROJDIR/plugins/check_procesors_configuration.py $EXPID

    if [ $? != 0 ] ; then
        get_out=true;
    fi
}
```
This is the original code of local_setup check_processors_configuration() and I noticed that it  will always load the default autosubmit version instead of the one that is being used

**Comment by kinow** [comment]
In GitLab by @mcastril on Dec 18, 2019, 18:37

Thanks @dbeltrankyl . From what you reported here, I see some actions (please correct if needed):

* [ ]   Ask Operations to change the Python environment with unloading / loading another Python module. Could you ask this @dbeltrankyl ?
* [ ]   Fix the config checker so it doesn't load AS default version, but the one it is currently being used. @pablohe

**Comment by kinow** [comment]
In GitLab by @pablohe on Jan 10, 2020, 18:02

Hi
About 

> Fix the config checker so it doesn't load AS default version, but the one it is currently being used. @pablohe

Do you know if the AS version is accessible from the templates using something like `%AS_version%`? 

I can't found it in https://autosubmit.readthedocs.io/en/latest/variables.html

If not I don't know how to get it (we are planning run it in all the platforms)

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 30, 2020, 18:14

We are coming again to this issue.

@dbeltrankyl so the problem with Autosubmit-Python3 was this line:

`module load bscearth.utils/0.5.2-foss-2015a-Python-2.7.9`

But as this line is only loaded to run Autosubmit then it will disappear when we use Autosubmit4.

**Comment by kinow** [comment]
In GitLab by @mcastril on Apr 30, 2020, 18:14

mentioned in merge request auto-ecearth3!248

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 13, 2021, 10:36

This issue will be retaken once the 3.13.0b version becomes the default version, we would need to update the 4.0 branch to match the new changes of 3.13.0.

**Comment by kinow** [comment]
In GitLab by @ggcr on Nov 9, 2022, 11:59

mentioned in issue autosubmit_api#12

**Comment by kinow** [comment]
In GitLab by @ggcr on Nov 10, 2022, 08:28

mentioned in issue autosubmit_api#14
